ufl/algorithms/comparison_checker.py

```diff
@@ -0,0 +1,125 @@
+from ufl.argument import Argument
+
+
+class CheckComparisons(MultiFunction):
+    """Raises an error if comparisons are done with complex quantities.
+
+    If quantities are real, adds the Real operator to the compared quantities.
+
+    Terminals that are real are RealValue, Zero, and Argument
+    (even in complex FEM, the basis functions are real)
+    Operations that produce reals are Abs, Real, Imag.
+    Terminals default to complex, and Sqrt, Pow (defensively) imply complex.
+    Otherwise, operators preserve the type of their operands.
+    """
+    def __init__(self):
+        MultiFunction.__init__(self)
+        self.nodetype = {}
+
+    def expr(self, o, *ops):
+        """Defaults expressions to complex unless they only
+        act on real quantities. Overridden for specific operators.
+
+        Rebuilds objects if necessary.
+        """
+
+        types = {self.nodetype[op] for op in ops}
+
+        if types:
+            t = "complex" if "complex" in types else "real"
+        else:
+            t = "complex"
+
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = t
+        return o
+
+    def compare(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("Ordering undefined for complex values.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    gt = compare
+    lt = compare
+    ge = compare
+    le = compare
+
+    def max_value(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with max.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def min_value(self, o, *ops):
+        types = {self.nodetype[op] for op in ops}
+
+        if "complex" in types:
+            raise ComplexComparisonError("You can't compare complex numbers with min.")
+        else:
+            o = o._ufl_expr_reconstruct_(*map(Real, ops))
+            self.nodetype[o] = "bool"
+            return o
+
+    def real(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'real'
+        return o
+
+    def imag(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'real'
+        return o
+
+    def sqrt(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        self.nodetype[o] = 'complex'
+        return o
+
+    def power(self, o, *ops):
+        o = self.reuse_if_untouched(o, *ops)
+        if float(ops[1]) < 1.0 and float(ops[1]) > 0.0:
+            self.nodetype[o] = 'complex'
+        elif self.nodetype[ops[0]] == 'complex':
+            self.nodetype[o] = 'complex'
+        else:
+            self.nodetype[o] = 'real'
+        return o
```

Comment:
I'm quite sceptical about the logic of this. (-2)^(3/2) is imaginary, but it seems function classifies it as real.